src/index.js

```diff
@@ -202,172 +286,266 @@ class PeerId {
-  if (!validMulticodec(cid)) throw new Error('Supplied PeerID CID has invalid multicodec: ' + cid.codec)
-  return new PeerIdWithIs(cid.multihash)
-}
+    if (typeof buf === 'string') {
+      buf = uint8ArrayFromString(String(key), 'base64pad')
+    }
 
-// Public Key input will be a Uint8Array
-exports.createFromPubKey = async (key) => {
-  let buf = key
+    if (!(buf instanceof Uint8Array)) {
+      throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
+    }
 
-  if (typeof buf === 'string') {
-    buf = uint8ArrayFromString(key, 'base64pad')
+    const pubKey = await cryptoKeys.unmarshalPublicKey(buf)
+    return computePeerId(pubKey)
   }
 
-  if (!(buf instanceof Uint8Array)) {
-    throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
-  }
+  /**
+   * Create PeerId from private key.
+   * 
+   * @param {Uint8Array | string} key - Private key, as Uint8Array or base64-encoded string.
+   * @returns {Promise<PeerId>}
+   */
+  static async createFromPrivKey (key) {
+    if (typeof key === 'string') {
+      key = uint8ArrayFromString(key, 'base64pad')
+    }
 
-  const pubKey = await cryptoKeys.unmarshalPublicKey(buf)
-  return computePeerId(null, pubKey)
-}
+    if (!(key instanceof Uint8Array)) {
+      throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
+    }
 
-// Private key input will be a string
-exports.createFromPrivKey = async (key) => {
-  if (typeof key === 'string') {
-    key = uint8ArrayFromString(key, 'base64pad')
+    const privKey = await cryptoKeys.unmarshalPrivateKey(key)
+    return computePeerId(privKey.public, privKey)
   }
 
-  if (!(key instanceof Uint8Array)) {
-    throw new Error('Supplied key is neither a base64 string nor a Uint8Array')
-  }
+  /**
+   * Create PeerId from PeerId JSON formatted object.
+   * 
+   * @param {PeerIdJSON} obj 
+   * @returns {Promise<PeerId>}
+   */
+  static async createFromJSON (obj) {
+    const id = mh.fromB58String(obj.id)
+    const rawPrivKey = obj.privKey && uint8ArrayFromString(obj.privKey, 'base64pad')
+    const rawPubKey = obj.pubKey && uint8ArrayFromString(obj.pubKey, 'base64pad')
+    const pub = rawPubKey && await cryptoKeys.unmarshalPublicKey(rawPubKey)
+
+    if (!rawPrivKey) {
+      return new PeerId(id, undefined, pub)
+    }
 
-  const privKey = await cryptoKeys.unmarshalPrivateKey(key)
-  return computePeerId(privKey, privKey.public)
-}
+    const privKey = await cryptoKeys.unmarshalPrivateKey(rawPrivKey)
+    const privDigest = await computeDigest(privKey.public)
 
-exports.createFromJSON = async (obj) => {
-  const id = mh.fromB58String(obj.id)
-  const rawPrivKey = obj.privKey && uint8ArrayFromString(obj.privKey, 'base64pad')
-  const rawPubKey = obj.pubKey && uint8ArrayFromString(obj.pubKey, 'base64pad')
-  const pub = rawPubKey && await cryptoKeys.unmarshalPublicKey(rawPubKey)
+    /**
+     * @type {Uint8Array}
+     */
+    let pubDigest = new Uint8Array()
```

Comment:
probably better to not allocate an uint8array here and "merge" the next if statements, which would make ts not complain I believe. What do you think?